### PR TITLE
Proof instructions and transaction model upgrade

### PR DESCRIPTION
**Pull request by iamyulong**
This is a mega PR that
- Added new proof instructions
- Renamed some existing instructions for consistency
- Cleaned up transaction builder to be consistent with transaction model
- Further reduced the usage of `ResourceSpecifier` (used by `parse_args()` only)
- Enhanced `Account` blueprint to support `withdraw`/`create_proof` by all, amount, or non-fungible ids.

# For new joiners
- All CALL returned resources go to the worktop
- All CALL returned proofs go to the auth zone (this is also where signature virtual proof sits)
- You can take resource from the worktop, which produces a `Bucket`
- You can take or derive proof from the auth zone, which produces a `Proof`
- Buckets can be passed to another component
- Proofs can be passed **once** to another component
- Proofs in the auth zone are automatically used for accessing component or the resource system

# Transaction manifest instructions

| Status  | Instruction                             | Argument 1                | Argument 2        | Argument 3 |
|---------|-----------------------------------------|---------------------------|-------------------|------------|
| Renamed  | `TAKE_FROM_WORKTOP`                     | `ResourceAddress`         |                   |            |
| Renamed  | `TAKE_FROM_WORKTOP_BY_AMOUNT`           | `Decimal`                 | `ResourceAddress` | `Bucket`   |
| Renamed  | `TAKE_FROM_WORKTOP_BY_IDS`              | `BTreeSet<NonFungibleId>` | `ResourceAddress` | `Bucket`   |
|         | `RETURN_TO_WORKTOP`                     | `Bucket`                  |                   |            |
| **New** | `ASSERT_WORKTOP_CONTAINS`                        | `ResourceAddress`         |                   |            |
| Renamed  | `ASSERT_WORKTOP_CONTAINS_BY_AMOUNT`              | `Decimal`                 | `ResourceAddress` |            |
| **New** | `ASSERT_WORKTOP_CONTAINS_BY_IDS`                 | `BTreeSet<NonFungibleId>` | `ResourceAddress` |            |
| Renamed  | `TAKE_FROM_AUTH_ZONE`                   | `Proof`                   |                   |            |
| Renamed  | `MOVE_TO_AUTH_ZONE`                      | `Proof`                   |                   |            |
| **New** | `CLEAR_AUTH_ZONE`                       |                           |                   |            |
| **New** | `CREATE_PROOF_FROM_AUTH_ZONE`           | `ResourceAddress`         | `Proof`           |            |
| **New** | `CREATE_PROOF_FROM_AUTH_ZONE_BY_AMOUNT` | `Decimal`                 | `ResourceAddress` | `Proof`    |
| **New** | `CREATE_PROOF_FROM_AUTH_ZONE_BY_IDS`    | `BTreeSet<NonFungibleId>` | `ResourceAddress` | `Proof`    |
| Renamed  | `CREATE_PROOF_FROM_BUCKET`              | `Bucket`                  | `Proof`           |            |
|         | `CLONE_PROOF`                           | `Proof`                   | `Proof`           |            |
|         | `DROP_PROOF`                            | `Proof`                   |                   |            |
|         | `CALL_FUNCTION`                         | `PackageAddress`          | `String`          | `String`   |
|         | `CALL_METHOD`                           | `ComponentAddress`        | `String`          |            |
|         | `CALL_METHOD_WITH_ALL_RESOURCES`        | `ComponentAddress`        | `String`          |            |
|         | `PUBLISH_PACKAGE`                       | `Blob`                    |                   |            |

# `Account` blueprint methods

| Name                     | Argument 1                | Argument 2        |
|--------------------------|---------------------------|-------------------|
| `withdraw`               | `ResourceAddress`         |                   |
| `withdraw_by_amount`     | `Decimal`                 | `ResourceAddress` |
| `withdraw_by_ids`        | `BTreeSet<NonFungibleId>` | `ResourceAddress` |
| `create_proof`           | `ResourceAddress`         |                   |
| `create_proof_by_amount` | `Decimal`                 | `ResourceAddress` |
| `create_proof_by_ids`    | `BTreeSet<NonFungibleId>` | `ResourceAddress` |
| `deposit`                | `Bucket`                  |                   |
| `deposit_batch`          | `Vec<Bucket>`             |                   |

### CODE REVIEW

File: assets/account/src/lib.rs [hunk 1]
```diff
@@ -88,5 +101,35 @@ blueprint! {
...
             }
         }
+
+        /// Create proof of resource.
+        pub fn create_proof_by_amount(
+            &self,
+            amount: Decimal,
+            resource_def_id: ResourceDefId,
+        ) -> Proof {
+            let vault = self.vaults.get(&resource_def_id);
```
Comment (talekhinezh) [inline assets/account/src/lib.rs:111]: Should we check for amount == 0 here? Should also add comment on what to expect if amount == 0

File: assets/account/src/lib.rs [hunk 2]
```diff
@@ -88,5 +101,35 @@ blueprint! {
...
+            }
+        }
+
+        /// Create proof of resource.
+        pub fn create_proof_by_ids(
+            &self,
+            ids: BTreeSet<NonFungibleId>,
+            resource_def_id: ResourceDefId,
+        ) -> Proof {
+            let vault = self.vaults.get(&resource_def_id);
```
Comment (talekhinezh) [inline assets/account/src/lib.rs:126]: Should we check for ids.is_empty() here? Should also add comment on what to expect if ids.is_empty()

**Review by talekhinezh — APPROVED** [review]

File: assets/account/src/lib.rs (hunk 1 above)
Comment (iamyulong) [reply assets/account/src/lib.rs:111]: Yes, I will make it clearer on what happens if `amount == zero`

File: assets/account/src/lib.rs (hunk 2 above)
Comment (iamyulong) [reply assets/account/src/lib.rs:126]: Yes

**Pull request merged**